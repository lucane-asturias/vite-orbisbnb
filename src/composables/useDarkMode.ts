import { ref } from 'vue'
import useNotification from './useNotification'

const darkModeActive = ref(false)

const useDarkMode = () => {
  const { setNotification } = useNotification()

  const toggleDarkMode = (): { active: boolean } => {
    darkModeActive.value = !darkModeActive.value

    const modeType = darkModeActive.value ? 'Dark Mode' : 'Light Mode'
    return setNotification(`${modeType} turned on!`)
  }

  return { 
    darkMode: darkModeActive,
    toggleDarkMode
  }
}

export default useDarkMode
