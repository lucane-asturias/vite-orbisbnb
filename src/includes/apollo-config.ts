import { ApolloClient, createHttpLink, InMemoryCache } from '@apollo/client/core';

// Declare the HTTP endpoint connection to GraphQL API
const httpLink = createHttpLink({
  uri: 'http://localhost:3000/api',
});

// Store the results of the API calls locally, in memory cache
const cache = new InMemoryCache();

const apolloClient = new ApolloClient({
  link: httpLink,
  cache
});

export default apolloClient