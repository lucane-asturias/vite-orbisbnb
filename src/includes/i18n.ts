import { createI18n } from 'vue-i18n'
/*
 * All i18n resources specified in the plugin `include` option can be loaded
 * at once using the import syntax
 */
import messages from '@intlify/unplugin-vue-i18n/messages'

export const i18n = createI18n({
  legacy: false,
  locale: 'en',
  fallbackLocale: 'pt',
  messages
})
