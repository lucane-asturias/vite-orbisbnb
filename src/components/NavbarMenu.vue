<script lang="ts" setup>
  import { computed } from 'vue'
  import { useI18n } from 'vue-i18n'
  import { setLocale } from '@vee-validate/i18n'
  import SearchListing from '@/modules/listings/components/SearchListing.vue'
  import { useListingsStore } from '@/modules/listings/store/listingsStore'
  import useDarkMode from '../composables/useDarkMode'

  const listingsStore = useListingsStore()
  const { darkMode, toggleDarkMode } = useDarkMode()

  const { locale } = useI18n({ useScope: 'global' })

  const darkModeButtonText = computed<string>(() => {
    switch (locale.value) {
      case 'pt':
        return darkMode.value ? 'Modo Claro' : 'Modo Escuro'
        break;
      case 'es':
        return darkMode.value ? 'Modo Luz' : 'Modo Oscuro'
        break;
      case 'ja':
        return darkMode.value ? 'ライトモード' : 'ダークモード'
        break;
      default:
        return darkMode.value ? 'Light Mode' : 'Dark Mode'
    }
  })

  const changeLocale = (event: PointerEvent) => {
    switch (event.target.value) {
      case 'pt':
        locale.value = 'pt'
        setLocale('pt_BR')
        break;
      case 'es':
        locale.value = 'es'
        setLocale('es')
        break;
      case 'ja':
        locale.value = 'ja'
        setLocale('ja')
        break;
      default:
        locale.value = 'en'
        setLocale('en')
    }
  }

</script>

<template>
  <nav class="border-gray-200 dark:bg-gray-900">
    <menu class="flex justify-between items-center mx-auto p-4 mx-5 lg:mx-7">

      <router-link to="/">
        <span class="self-center text-md md:text-xl font-semibold whitespace-nowrap dark:text-white">
          &#9992; {{ $t('navbar_menu.orbis_bnb') }} &#9992;
        </span>
      </router-link>

      <div class="flex items-center">
        <SearchListing :darkMode="darkMode" />
        <button 
          class="hidden sm:block ml-5 text-md font-semibold py-1 sm:py-3 px-3 rounded shadow float-right" 
          :class="{ 
           'text-gray-100 bg-slate-800 hover:bg-gray-900': darkMode,
           'text-gray-900 bg-slate-200 hover:bg-gray-300': !darkMode
          }"
          @click="toggleDarkMode"
        >
         {{ darkModeButtonText }}
        </button>
        <select @click.prevent="changeLocale"
          class="fa cursor-pointer appearance-none text-lg w-11 sm:w-12 h-full sm:h-12 bg-gray-200 border border-gray-200 text-gray-500 
            py-2.5 sm:py-2.5 px-3 sm:px-3.5 ml-5 rounded leading-tight focus:outline-none focus:border-gray-500 font-bold tracking-normal lg:tracking-widest"
          :class="{ 
            'bg-slate-800 hover:bg-lightgreyish focus:bg-slate-800': darkMode, 
            'bg-gray-200 hover:bg-gray-300': !darkMode 
          }"

        >
          <option hidden selected value="#" aria-hidden="true">&#xf57d;</option>
          <option disabled>{{ $t('navbar_menu.select_language') }}</option>
          <option value="en">{{ $t('language.en') }}</option>
          <option value="pt">{{ $t('language.pt') }}</option>
          <option value="es">{{ $t('language.es') }}</option>
          <option value="ja">{{ $t('language.ja') }}</option>
        </select>

      </div>
      <div id="darkmodetoggle" @click="toggleDarkMode"
        class="block sm:hidden darkmode-fixed-bottom-left"
        :class="darkMode ? 'darkmodetoggle' : ''">
        Toggle darkmode
      </div>
    </menu>
  </nav>
</template>

<style lang="css" scoped>
.darkmode-fixed-bottom-left {
  position: fixed !important;
  z-index: 9999999999;
  bottom: 50px !important;
  top: unset;
  left: 37px;
  /*  
    fixed z-90 bottom-10 inset-x-0 left-12
    cursor-pointer bg-black w-8 h-8 rounded-full
  */
}

#darkmodetoggle {
  cursor: pointer;
  background: #000;
  background: linear-gradient(225deg,#000 50%,#fff 50%);
  width: 33px;
  height: 33px;
  border-radius: 100%;
  border: 1px solid #000;
  font-size: 0; /* remove innertext */
}

#darkmodetoggle:before {
  content: "Enable Darkmode";
  position: absolute;
  top: 50%;
  transform: translatey(-50%);
  left: 100%;
  margin-left: 15px;
  width: 110px;
  padding: 5px 2px;
  border-radius: 10px;
  background: #000;
  color: #fff;
  text-align: center;
  font-size: 11px !important;
  display: none;
}

#darkmodetoggle:hover:before {
  display: block;
  background-color: black;
  color: whitesmoke;
  font-weight: 600;
}

.darkmodetoggle {
  background: linear-gradient(45deg,#000 50%,#fff 50%) !important;
  border: 1px solid #fff;
}

.darkmodetoggle:before {
  content: "Disable Darkmode" !important;
  background: #fff;
  color: #000;
}

.darkmodetoggle:hover:before {
  display: block;
  background-color: whitesmoke !important;
  color: black !important;
  font-weight: 600;
}
</style>