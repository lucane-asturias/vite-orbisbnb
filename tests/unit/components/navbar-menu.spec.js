import { shallowMount, RouterLinkStub } from '@vue/test-utils'
import { createPinia, setActivePinia } from 'pinia'
import NavbarMenu from '@/components/NavbarMenu.vue'
import useDarkMode from '@/composables/useDarkMode'
import { i18n } from '@/includes/i18n'

describe('Tests on the NavbarMenu.vue component', () => {

  let wrapper
  // SETUP - run prior to each unit test
  beforeEach(() => {
    // creates a fresh pinia and make it active so it's automatically picked
    // up by any useStore() call without having to pass it to it:
    // `useStore(pinia)`
    setActivePinia(createPinia())
    // render the component
    wrapper = shallowMount(NavbarMenu, {
      global: {
        components: { 'router-link': RouterLinkStub },
        mocks: { $t: (translation) => translation },
        plugins: [i18n],
      }
    })
  })

  test('should match snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('click event should toggle darkmode', () => {
    expect(wrapper.vm.darkMode).toBeFalsy()
    wrapper.find('#darkmodetoggle').trigger('click')
    expect(wrapper.vm.darkMode).toBeTruthy()
  })

  test('orbisbnb routerlink should lead to the root page', () => {
    expect(wrapper.findComponent(RouterLinkStub).props().to).toMatch('/')
  })

  test('darkmode computed fc should return the english version', () => {
    expect(wrapper.vm.darkModeButtonText).toMatch('Light Mode')
    wrapper.find('#darkmodetoggle').trigger('click')
    expect(wrapper.vm.darkModeButtonText).toMatch('Dark Mode')
  })

  test('darkmode computed fc should return other language', async () => {
    wrapper.vm.locale = 'ja'
    expect(wrapper.vm.darkModeButtonText).toMatch('ダークモード')
    wrapper.find('#darkmodetoggle').trigger('click')
    expect(wrapper.vm.darkModeButtonText).toMatch('ライトモード')
  })

  test('the button for the computed darkModeButtonText shouldve dynamic classes', async () => {
    expect(wrapper.vm.darkMode).toBeTruthy()
    expect(wrapper.find('button').classes('text-gray-100 bg-slate-800 hover:bg-gray-900'))
    wrapper.find('#darkmodetoggle').trigger('click')
    expect(wrapper.vm.darkMode).toBeFalsy()
    expect(wrapper.find('button').classes('text-gray-900 bg-slate-200 hover:bg-gray-300'))
  })

  test('select language button should be disabled by default', () => {
    expect(wrapper.findAll('option').at(1).element.disabled).toBeTruthy()
  })

})