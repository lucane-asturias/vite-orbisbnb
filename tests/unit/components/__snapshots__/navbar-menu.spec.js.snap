// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tests on the NavbarMenu.vue component > should match snapshot 1`] = `
"<nav data-v-93d49201=\\"\\" class=\\"border-gray-200 dark:bg-gray-900\\">
  <menu data-v-93d49201=\\"\\" class=\\"flex justify-between items-center mx-auto p-4 mx-5 lg:mx-7\\">
    <router-link-stub data-v-93d49201=\\"\\" to=\\"/\\" custom=\\"false\\"></router-link-stub>
    <div data-v-93d49201=\\"\\" class=\\"flex items-center\\">
      <search-listing-stub data-v-93d49201=\\"\\" darkmode=\\"false\\"></search-listing-stub><button data-v-93d49201=\\"\\" class=\\"hidden sm:block ml-5 text-md font-semibold py-1 sm:py-3 px-3 rounded shadow float-right text-gray-900 bg-slate-200 hover:bg-gray-300\\">Dark Mode</button><select data-v-93d49201=\\"\\" class=\\"fa cursor-pointer appearance-none text-lg w-11 sm:w-12 h-full sm:h-12 bg-gray-200 border border-gray-200 text-gray-500 py-2.5 sm:py-2.5 px-3 sm:px-3.5 ml-5 rounded leading-tight focus:outline-none focus:border-gray-500 font-bold tracking-normal lg:tracking-widest bg-gray-200 hover:bg-gray-300\\">
        <option data-v-93d49201=\\"\\" hidden=\\"\\" aria-hidden=\\"true\\" value=\\"#\\">ï•½</option>
        <option data-v-93d49201=\\"\\" disabled=\\"\\">navbar_menu.select_language</option>
        <option data-v-93d49201=\\"\\" value=\\"en\\">language.en</option>
        <option data-v-93d49201=\\"\\" value=\\"pt\\">language.pt</option>
        <option data-v-93d49201=\\"\\" value=\\"es\\">language.es</option>
        <option data-v-93d49201=\\"\\" value=\\"ja\\">language.ja</option>
      </select>
    </div>
    <div data-v-93d49201=\\"\\" id=\\"darkmodetoggle\\" class=\\"block sm:hidden darkmode-fixed-bottom-left\\"> Toggle darkmode </div>
  </menu>
</nav>"
`;
